File: factory/ExerciseFactory.java
package factory;

import model.Exercise;
import model.SentenceExercise;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ExerciseFactory {

    private static final int EXERCISES_PER_SESSION = 8;

    private static List<String> loadPhrases(String filePath) {

        try {
            return Files.readAllLines(
                    Paths.get(filePath),
                    StandardCharsets.UTF_8
            );
        }
        catch (IOException e) {
            System.out.println("Erro ao ler arquivo: " + filePath);
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    private static List<Exercise> createExercisesFromFile(
            String filePath,
            String category,
            int difficulty
    ) {

        List<String> phrases = loadPhrases(filePath);

        Collections.shuffle(phrases);

        List<Exercise> exercises = new ArrayList<>();

        int amount = Math.min(
                EXERCISES_PER_SESSION,
                phrases.size()
        );

        for (int i = 0; i < amount; i++) {

            exercises.add(
                    new SentenceExercise(
                            phrases.get(i),
                            "Pratique a digitação",
                            category,
                            difficulty
                    )
            );
        }

        return exercises;
    }

    private static List<Exercise> createLevel1() {
        return createExercisesFromFile(
                "assets/exercises/1-minusculas.txt",
                "Minúsculas",
                1
        );
    }

    private static List<Exercise> createLevel2() {
        return createExercisesFromFile(
                "assets/exercises/2-maiusculas.txt",
                "Maiúsculas",
                2
        );
    }

    private static List<Exercise> createLevel3() {
        return createExercisesFromFile(
                "assets/exercises/3-numeros.txt",
                "Números",
                3
        );
    }

    private static List<Exercise> createLevel4() {
        return createExercisesFromFile(
                "assets/exercises/4-acentos.txt",
                "Acentos",
                4
        );
    }

    private static List<Exercise> createLevel5() {
        return createExercisesFromFile(
                "assets/exercises/5-pontuacao.txt",
                "Pontuação",
                5
        );
    }

    public static List<Exercise> createExercisesForLevel(int level) {

        switch (level) {

            case 1:
                return createLevel1();

            case 2:
                return createLevel2();

            case 3:
                return createLevel3();

            case 4:
                return createLevel4();

            case 5:
                return createLevel5();

            default:
                return createLevel1();
        }
    }

    public static List<Exercise> createCampaign() {

        List<Exercise> campaign = new ArrayList<>();

        campaign.addAll(createLevel1());
        campaign.addAll(createLevel2());
        campaign.addAll(createLevel3());
        campaign.addAll(createLevel4());
        campaign.addAll(createLevel5());

        return campaign;
    }
}

File: backend/src/game/exerciseFactory.js
// Porta de factory/ExerciseFactory.java — mesma lógica de leitura, embaralhamento
// e montagem da campanha (5 módulos x 8 frases, todas como "SentenceExercise").

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.join(__dirname, "..", "..", "data", "exercises");

const EXERCISES_PER_SESSION = 8;

const MODULES = [
  { file: "1-minusculas.txt", category: "Minúsculas", difficulty: 1 },
  { file: "2-maiusculas.txt", category: "Maiúsculas", difficulty: 2 },
  { file: "3-numeros.txt", category: "Números", difficulty: 3 },
  { file: "4-acentos.txt", category: "Acentos", difficulty: 4 },
  { file: "5-pontuacao.txt", category: "Pontuação", difficulty: 5 },
];

function loadPhrases(fileName) {
  const filePath = path.join(DATA_DIR, fileName);
  try {
    return fs
      .readFileSync(filePath, "utf-8")
      .split(/\r?\n/)
      .filter((line) => line.length > 0);
  } catch (err) {
    console.error(`Erro ao ler arquivo: ${filePath}`, err);
    return [];
  }
}

function shuffle(array) {
  const arr = [...array];
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

// Mesma fórmula de SentenceExercise.java:
// xpReward = 60 * difficulty; timeLimit = 120 + difficulty * 42
function buildSentenceExercise(sentence, category, difficulty, moduleMeta) {
  return {
    targetText: sentence,
    description: "Pratique a digitação",
    category,
    xpReward: 60 * difficulty,
    difficulty,
    timeLimit: 120 + difficulty * 42,
    instructions: "Digite a frase completa. Use espaço entre as palavras.",
    ...moduleMeta,
  };
}

function createLevel({ file, category, difficulty }, moduleIndex, totalModules) {
  const phrases = shuffle(loadPhrases(file));
  const amount = Math.min(EXERCISES_PER_SESSION, phrases.length);
  const exercises = [];
  for (let i = 0; i < amount; i++) {
    exercises.push(
      buildSentenceExercise(phrases[i], category, difficulty, {
        moduleNumber: difficulty, // 1..5, número absoluto do módulo (= nível de dificuldade)
        moduleIndex, // posição do módulo NESTA campanha (1-based)
        totalModules, // quantos módulos tem esta campanha
        indexInModule: i + 1,
        totalInModule: amount,
      })
    );
  }
  return exercises;
}

// startLevel (1 a 5): a campanha começa nesse módulo e segue até o 5º,
// igual descrito no README ("a campanha segue dele até o final").
export function createCampaign(startLevel = 1) {
  const safeStart = Math.min(Math.max(1, Math.trunc(startLevel) || 1), MODULES.length);
  const selected = MODULES.slice(safeStart - 1);
  const totalModules = selected.length;
  return selected.flatMap((mod, i) => createLevel(mod, i + 1, totalModules));
}

export function getModules() {
  return MODULES.map(({ category, difficulty }) => ({ moduleNumber: difficulty, category }));
}
